doc/sphinx/source/howto/derive-for-training.rst

```diff
@@ -0,0 +1,140 @@
+variables. Note that this is not the same as *constructing* the training model 
+directly, which would be significantly more work.   
+
+
+.. _code_structure:
+
+Code Structure
+==============
+
+
+.. _inference:
+
+Inference
+---------
+
+We begin by building the graph, starting with the input parameter 
+``X``. We define a fully-connected layer, including a parameter for
+weights and bias.
+
+.. literalinclude:: ../../../examples/mnist_mlp.cpp
+   :language: cpp
+   :lines: 123-135
+
+
+We repeat the process for the next layer, which we
+normalize with a ``softmax``.
+
+.. literalinclude:: ../../../examples/mnist_mlp.cpp
+   :language: cpp
+   :lines: 137-148
+
+
+.. _loss:
+
+Loss
+----
+
+We use cross-entropy to compute the loss. nGraph does not currenty
+have a cross-entropy op, so we implement it directly, adding clipping
+to prevent underflow.
+
+.. literalinclude:: ../../../examples/mnist_mlp.cpp
+   :language: cpp
+   :lines: 150-164
+
+
+.. _backprop:
+
+Backprop
+--------
+
+To compute the updates, we need a computation that computes an
+adjustment for the weights from an adjustment to the loss. In nGraph,
+``loss`` is the computation that computes the loss; it is equivalent
+to what some descriptions of the autodiff algorithm call "the tape."
+If each step of a computation between a weight and the loss has a
+derivative, we can use the reverse mode autodiff to compute an update
+for the weight from an update for the loss; in fact, we can compute
+updates for all the weights, sharing much of the update computation,
+and this is what some frameworks do. But it is just as easy for us to
+instead create the update computations for all of the weights, which
+lets compilation optimize across steps in the computation.
+
+We'll call the adjustment to the loss
+
+.. code-block:: cpp
+
+   auto delta = -learning_rate * loss;
+
+For any node ``N``, if the update for ``loss`` is ``delta``, the
+update computation for ``N`` will be given by the node
+
+.. code-block:: cpp
+
+   auto update = loss->backprop_node(N, delta);
+
+The different update nodes will share intermediate computations. So to
+get the updated value for the weights we just say:
+
+.. literalinclude:: ../../../examples/mnist_mlp.cpp
+   :language: cpp
+   :lines: 166-176
+
+.. _update:
+
+Update
+------
+
+Since nGraph is stateless, we train by making a function that has the
+original weights among its inputs and the updated weights among the
+results. For training, we'll also need the labeled training data as
+inputs, and we'll return the loss as an additional result.  We'll also
+want to track how well we are doing; this is a function that returns
+the loss and has the labeled testing data as input. Although we can
+use the same nodes in different functions, nGraph currently does not
+allow the same nodes to be compiled in different functions, so we
+compile clones of the nodes.
+
+.. literalinclude:: ../../../examples/mnist_mlp.cpp
+   :language: cpp
+   :lines: 246-253
+
```

Comment:
Some example output might be good at the end here. Also, are we giving people the information they need to prepare the dataset? (Maybe that's in a different file)